File: code/server/utility/Logger.cs
// Copyright (C) NOMAD Group <nomad-group.net>.
// For licensing information see LICENSE at the root of this distribution.

using System;
using System.Collections.Generic;

namespace noda
{
    [System.Flags]
    public enum LogLevel
    {
        Trace,    //< very detailed and extensive debugging information
        Debug,    //< less detailed debug information
        Info,     //< status information
        Warning,  //< minor or potential problems
        Error,    //< major problems that block further execution of a task
        Critical, //< major problems that block continued execution of everything
        Count     //< total number of log levels
    }

    public interface ILogSink
    {
        public void Write(string message, LogLevel level);
    }

    public class Logger
    {
        private static readonly string TimeFormat = "[HH:mm]";
        public List<ILogSink> Sinks;

        public Logger()
        {
            Sinks = new List<ILogSink>();
        }

        public void Error(string text)
        {
            Log(LogLevel.Error, text);
        }

        public void Critical(string text)
        {
            Log(LogLevel.Critical, text);
        }

        public void Info(string text)
        {
            Log(LogLevel.Info, text);
        }

        public void Trace(string text)
        {
            Log(LogLevel.Trace, text);
        }

        public void Warning(string text)
        {
            Log(LogLevel.Warning, text);
        }

        public void Log(LogLevel level, string text)
        {
            foreach (ILogSink sink in Sinks)
            {
                sink.Write(text, level);
            }
        }

        public static string FormatLogMessage(LogLevel level, string text)
        {
            string prefix = DateTime.Now.ToString(TimeFormat) + " <" +  level.ToString() + "> ";
            return prefix + text;
        }
    }
}

File: code/client/utils/Logger.cpp
// Copyright (C) NOMAD Group <nomad-group.net>.
// For licensing information see LICENSE at the root of this distribution.

#include "IdaInc.h"
#include "Logger.h"

namespace noda::utils
{
  static const char *GetLevelName(LogLevel log_level)
  {
#define LVL(x)      \
  case LogLevel::x: \
	return #x
	switch(log_level) {
	  LVL(Trace);
	  LVL(Info);
	  LVL(Warning);
	  LVL(Error);
	default:
	  return "???";
	}
#undef LVL
  }

  // pretty print me, please :-)
  void PrintLogMessageImpl(LogLevel ll, const char *text, const fmt::format_args &args)
  {
	const char *levelName = GetLevelName(ll);

	auto fmt = fmt::format("[Noda] <{}>: {}\n", levelName, fmt::vformat(text, args));
	callui(ui_msg, fmt.c_str());
  }
} // namespace noda::utils